### PR TITLE
Make joinPossiblyNormalCompletionWithAbruptCompletion live up to its advertised behavior

### PR DESCRIPTION
**Pull request by hermanventer**
Release note: none

joinPossiblyNormalCompletionWithAbruptCompletion said that it always returned effects with completions of type JoinedAbruptCompletions. Some callers took it at is word. Turns out it was lying.

Fixed things so that the advertising is correct. Also fixed an issue with a generator being lost during joining. (This was a latent bug uncovered by this fix.)

Refactored the code that fails when a loop body contains both break and continue completions governed by abstract conditions.

### CODE REVIEW

File: src/methods/join.js [hunk 1]
```diff
@@ -700,9 +726,9 @@ export class JoinImplementation {
 
   composeGenerators(realm: Realm, generator1: Generator, generator2: Generator): Generator {
     let result = new Generator(realm, "composed");
-    if (!generator1.empty() || !generator2.empty()) {
-      result.composeGenerators(generator1, generator2);
-    }
+    // We copy the entries here because actually composing the generators breaks the serializer
```
Comment (NTillmann) [inline src/methods/join.js]: Why?

**Comment by NTillmann** [comment]
If this fixes something, can we get a regression test case?

**Comment by hermanventer** [comment]
This is a refactor on the way towards fixing something. Lots of existing test cases touch this code.

File: src/methods/join.js (hunk 1 above)
Comment (hermanventer) [reply src/methods/join.js]: It seems like a bug or short coming in the serializer. I do not quite understand why.

Comment (NTillmann) [reply src/methods/join.js]: That is concerning; maybe hiding something. After landing this, can you file an issue with steps to reproduce?
(E.g. the old code to plug in, and which test case fails.)

**Pull request closed without merging**